# Sprocket Central Pty Ltd Data Set

### Importing necessary libraries

In [1]:
import numpy as np
import statistics as st
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from pandas import ExcelWriter
from pandas import ExcelFile

#### Creating the dataframe

In [2]:
data = pd.read_excel(r'C:\Users\Dell\Downloads\KPMG.xlsx')

#### Creating separate dataframe for each sheets

In [3]:
Transaction = pd.read_excel(r'C:\Users\Dell\Downloads\KPMG.xlsx',sheet_name = 1,skiprows = 1)
CustomerDemographic = pd.read_excel(r'C:\Users\Dell\Downloads\KPMG.xlsx', sheet_name = 3 , skiprows = 1)
CustomerAddress = pd.read_excel(r'C:\Users\Dell\Downloads\KPMG.xlsx', sheet_name = 4, skiprows = 1)

## Transaction Dataset

In [4]:
Transaction.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [5]:
Transaction.shape

(20000, 13)

In [6]:
Transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [7]:
Transaction.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

### Checking the null values of Transaction data

In [8]:
Transaction.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

#### checking if the null values of column brand and other 197 null values column are of same rows

In [ ]:
null_check = Transaction[Transaction['brand'].isnull() & Transaction['product_line'].isnull() & Transaction['product_class'].isnull() & Transaction['product_size'].isnull() & Transaction['standard_cost'].isnull() & Transaction['product_first_sold_date'].isnull()]

In [ ]:
Transaction.brand.value_counts()

In [11]:
null_check.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order                 2
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

Hence, 
    [brand, product_line, product_class, product_size, standard_cost, product_first_sold_date]
    These 6 columns have same number of null values so 197 in same rows so instead of imputing these many values for each column, it is better to drop these null values. As these rows are having multiple null values for many columns.

#### Dropping null values of brand in Transaction dataframe

In [12]:
transaction_1 = Transaction[Transaction['brand'].isnull()]

In [13]:
Transaction.drop(transaction_1.index,inplace = True)

In [14]:
# Rechecking the null values
Transaction.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               358
order_status                 0
brand                        0
product_line                 0
product_class                0
product_size                 0
list_price                   0
standard_cost                0
product_first_sold_date      0
dtype: int64

#### taking care of null values of online_order

In [15]:
# finding mode value
mode_online_order = Transaction.online_order.mode()[0]

In [16]:
# filling null values with mode
Transaction['online_order'].fillna(mode_online_order, inplace = True)

### Rechecking of null values

In [17]:
Transaction.isnull().sum()

transaction_id             0
product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

#### Making of a new column namely Profit which will be helpful in later stages

In [18]:
Transaction.profit = Transaction['list_price'] - Transaction['standard_cost']

### Checking the spellings for object dtype column

In [19]:
Transaction.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19803 non-null  float64       
 5   order_status             19803 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [20]:
Transaction.order_status.unique()

array(['Approved', 'Cancelled'], dtype=object)

In [21]:
Transaction.brand.unique()

array(['Solex', 'Trek Bicycles', 'OHM Cycles', 'Norco Bicycles',
       'Giant Bicycles', 'WeareA2B'], dtype=object)

In [22]:
Transaction.product_line.unique()

array(['Standard', 'Road', 'Mountain', 'Touring'], dtype=object)

In [23]:
Transaction.product_class.unique()

array(['medium', 'low', 'high'], dtype=object)

In [24]:
Transaction.product_size.unique()

array(['medium', 'large', 'small'], dtype=object)

##### Hence, Everything is ready for Transaction table.

#### ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## CustomerDemographic Dataset

In [25]:
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [26]:
CustomerDemographic.shape

(4000, 13)

In [27]:
CustomerDemographic.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

In [28]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [29]:
CustomerDemographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

### Checking null values count for CustomerDemographic

In [30]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

### Let's remove Default from data set

In [31]:
CustomerDemographic.drop('default',axis = 1,inplace = True)

In [32]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

### Examining  the deceased_indicator column

In [33]:
CustomerDemographic.deceased_indicator.value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [34]:
# Create a data frame for deceased customer
Deceased_I = CustomerDemographic[CustomerDemographic['deceased_indicator']== 'Y']
Deceased_I.head()

customer_id first_name    last_name  gender  \
752           753       Josy  St. Quentin  Female   
3789         3790     Kurtis       Morson    Male   

      past_3_years_bike_related_purchases        DOB      job_title  \
752                                    82 1970-03-07   Food Chemist   
3789                                   91 1959-05-31  Senior Editor   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
752                 Health  Affluent Customer                  Y      Yes   
3789                Retail      Mass Customer                  Y      Yes   

      tenure  
752      6.0  
3789     9.0

#### Dropping deceased index for 'Y' 


In [35]:
CustomerDemographic.drop(Deceased_I.index,axis = 0,inplace = True)

In [36]:
# checking value_count for deceased indicator
CustomerDemographic.deceased_indicator.value_counts()

N    3998
Name: deceased_indicator, dtype: int64

### Lets check the job_industry_category Column

In [37]:
CustomerDemographic.job_industry_category.isnull().sum()

656

In [38]:
CustomerDemographic.job_industry_category.value_counts()

Manufacturing         799
Financial Services    774
Health                601
Retail                357
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [39]:
CustomerDemographic.job_industry_category.value_counts(normalize = True)*100

Manufacturing         23.907840
Financial Services    23.159785
Health                17.983244
Retail                10.682226
Property               7.989228
IT                     6.672651
Entertainment          4.069420
Argiculture            3.381209
Telecommunications     2.154399
Name: job_industry_category, dtype: float64

#### Here, we need to foucus on top for data values
1. Manufacturing         23.907840
2. Financial Services    23.159785
3. Health                17.983244
4. Retail                10.682226

#### Hence, we can fill  the data with these values  

#### Lets check the percentage of null value in job_industry_category

In [40]:
CustomerDemographic.job_industry_category.isnull().mean()*100

16.408204102051023

#### As per value count and percentage of the missing value we can use following method :
1. As the percentage for  Manufacturing  is 23.907840 and is Financial Services  23.159785 
2. So we can fill 246 rows with Manufacturing and other 246 with Financial Services   
3. Remaining 164 we can split the half with Health and Retail

#### Lets create a new data frame for null rows of job_industry_category

In [41]:
Customer_null =  CustomerDemographic[CustomerDemographic.job_industry_category.isnull()]

In [42]:
#### Lets check the data frame Customer_null

Customer_null.isnull().mean()*100

customer_id                              0.000000
first_name                               0.000000
last_name                                3.810976
gender                                   0.000000
past_3_years_bike_related_purchases      0.000000
DOB                                      0.000000
job_title                               16.006098
job_industry_category                  100.000000
wealth_segment                           0.000000
deceased_indicator                       0.000000
owns_car                                 0.000000
tenure                                   0.000000
dtype: float64

#### Lets  fill the data with our approach

In [43]:
# creating a list to fill
fill_values = ['Manufacturing'] * 246 + ['Financial Services'] * 246 + ['Health'] * 82 + ['Retail'] * 82

In [44]:
# filling the null values for job_industry_category

Customer_null['job_industry_category'].fillna(value=pd.Series(fill_values, index=Customer_null.index), inplace=True)

In [45]:
# rechecking the null valuesof customer_null
Customer_null.isnull().sum()

customer_id                              0
first_name                               0
last_name                               25
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                              105
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
dtype: int64

#### now imputing nullvalues for job_industry_category

In [46]:
# deleting customer_null values from dataframe
CustomerDemographic.drop(Customer_null.index, axis = 0, inplace = True)

In [47]:
# checking shape
CustomerDemographic.shape

(3342, 12)

In [48]:
# appending with customer_null values
CustomerDemographic = CustomerDemographic.append(Customer_null)

In [49]:
# checking shape
CustomerDemographic.shape

(3998, 12)

### Rechecking null values

In [50]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

### dealing with tenure null values

In [51]:
CustomerDemographic.tenure.value_counts()

7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
14.0    200
9.0     199
6.0     191
4.0     191
13.0    191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: tenure, dtype: int64

In [52]:
CustomerDemographic.tenure.mode()

0    7.0
Name: tenure, dtype: float64

In [53]:
CustomerDemographic.tenure.median()

11.0

##### As multiple values are there near mode values, hence taking median as the means of filling null values for Tenure column. 

In [54]:
tenure_median = CustomerDemographic['tenure'].median()
CustomerDemographic['tenure'].fillna(tenure_median,inplace = True)

### Rechecking of null values

In [55]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
dtype: int64

#### Leaving the rest null_values as it is -As these columns will be treated again later after merging 

### Checking the spelling mistakes

##### Checking for gender ####

In [56]:
CustomerDemographic.gender.unique()

array(['F', 'Male', 'Female', 'U', 'Femal', 'M'], dtype=object)

In [57]:
# correcting the spellings
CustomerDemographic['gender'] = CustomerDemographic['gender'].map(lambda x : 'Female' if x == 'F' or x == "Femal" else ('Male' if x == 'M' else ( "Unspecified" if x == 'U' else x)))

In [58]:
# Rechecking the value_counts

CustomerDemographic['gender'].value_counts()

Female         2038
Male           1872
Unspecified      88
Name: gender, dtype: int64

In [59]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3998 entries, 0 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3998 non-null   int64         
 1   first_name                           3998 non-null   object        
 2   last_name                            3873 non-null   object        
 3   gender                               3998 non-null   object        
 4   past_3_years_bike_related_purchases  3998 non-null   int64         
 5   DOB                                  3911 non-null   datetime64[ns]
 6   job_title                            3492 non-null   object        
 7   job_industry_category                3998 non-null   object        
 8   wealth_segment                       3998 non-null   object        
 9   deceased_indicator                   3998 non-null   object        
 10  owns_car    

#### Checking for job_industry_category ####

In [60]:
CustomerDemographic.job_industry_category.unique()

array(['Health', 'Financial Services', 'Property', 'IT', 'Retail',
       'Argiculture', 'Manufacturing', 'Telecommunications',
       'Entertainment'], dtype=object)

In [61]:
# correcting the spelling of Agriculture
CustomerDemographic['job_industry_category'] = CustomerDemographic['job_industry_category'].replace('Argiculture','Agriculture')

#### Checking for wealth_segment ####

In [62]:
CustomerDemographic.wealth_segment.unique()

array(['Mass Customer', 'High Net Worth', 'Affluent Customer'],
      dtype=object)

#### Checking for owns_car ####

In [63]:
CustomerDemographic.owns_car.unique()

array(['Yes', 'No'], dtype=object)

### Adding the new column 'Age'

In [64]:
CustomerDemographic['DOB'] = pd.to_datetime(CustomerDemographic['DOB'])
CustomerDemographic['Age'] = 2017 - CustomerDemographic['DOB'].dt.year
CustomerDemographic['Age']

0       64.0
1       37.0
2       63.0
3       56.0
5       51.0
        ... 
3965    49.0
3973    56.0
3975    60.0
3982    52.0
3999    26.0
Name: Age, Length: 3998, dtype: float64

#### -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Customer Address Dataset

### Going through the dataset

In [65]:
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [66]:
CustomerAddress.shape

(3999, 6)

In [67]:
CustomerAddress.columns

Index(['customer_id', 'address', 'postcode', 'state', 'country',
       'property_valuation'],
      dtype='object')

In [68]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [69]:
CustomerAddress.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

### Checking for null value counts

In [70]:
CustomerAddress.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

### Checking for spelling mistakes

In [71]:
# checking with address
CustomerAddress.address.unique()

array(['060 Morning Avenue', '6 Meadow Vale Court', '0 Holy Cross Court',
       ..., '87 Crescent Oaks Alley', '8194 Lien Street',
       '320 Acker Drive'], dtype=object)

In [72]:
# checking with country
CustomerAddress.country.unique()

array(['Australia'], dtype=object)

In [73]:
# checking with states
CustomerAddress.state.unique()

array(['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria'], dtype=object)

In [74]:
# replacing the values with correct one

CustomerAddress['state'].replace('QLD','Queensland',inplace = True)
CustomerAddress['state'].replace('VIC','Victoria',inplace = True)
CustomerAddress['state'].replace('NSW','New South Wales',inplace = True)

In [75]:
# rechecking with states
CustomerAddress.state.unique()

array(['New South Wales', 'Queensland', 'Victoria'], dtype=object)

### ----------------------------------------------------------------------------------------------------------------------------------------------------------------

## NewCustomer Dataset

In [77]:
# reading the data
New_customer = pd.read_excel(r'C:\Users\Dell\Downloads\KPMG.xlsx',sheet_name = 2,skiprows = 1)

### Going through the data

In [78]:
New_customer.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.56      0.7000       0.8750     0.743750   
1                  11        0.89      0.8900       1.1125     0.945625   
2                   5        1.01      1.0100       1.0100     1.010000   
3                   1        0.87      1.0875       1.0875     1.087500   
4                   9        0.52      0.5200       0.6500     0.650000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [79]:
New_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [80]:
New_customer.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20', 'Rank', 'Value'],
      dtype='object')

In [81]:
New_customer.shape

(1000, 23)

### Dealing with the null values

In [82]:
New_customer.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Unnamed: 16                              0
Unnamed: 17                              0
Unnamed: 18                              0
Unnamed: 19                              0
Unnamed: 20                              0
Rank                                     0
Value                                    0
dtype: int6

#### Dealing with job_industry_category null values

In [83]:
# checking the value_counts for job_industry_category
New_customer.job_industry_category.value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

Now using the same approach as we used previously to tackle job_industry_category column null values. 
Filling the null values with top 4 repeating categories of job_industry_category and substanially distributing them against the null values

In [85]:
# fillvalues list for job_industry_category null values
fill_values_new_customer = ['Manufacturing'] * 62 +['Financial Services'] * 61 +['Health'] * 21 +['Retail'] * 21

In [87]:
# creating a seperate dataframe for job_industry_category null values
Null_for_job_industry_category = New_customer[New_customer['job_industry_category'].isnull()]

In [88]:
# filling values in Null_for_job_industry_category by filling values list
Null_for_job_industry_category['job_industry_category'].fillna(value = pd.Series(fill_values_new_customer, index = Null_for_job_industry_category.index), inplace = True)

In [90]:
# dropping the null values of job_industry_category by referring from Null_for_job_industry_category
New_customer.drop(Null_for_job_industry_category.index , inplace = True)

In [91]:
# appending the null values for job_industry_category by refering to Null_for_job_industry_category
New_customer = New_customer.append(Null_for_job_industry_category)

In [93]:
# checking the shape
New_customer.shape

(1000, 23)

#### rechecking the null values

In [94]:
New_customer.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Unnamed: 16                              0
Unnamed: 17                              0
Unnamed: 18                              0
Unnamed: 19                              0
Unnamed: 20                              0
Rank                                     0
Value                                    0
dtype: int6

### Checking the spelling mistakes

In [95]:
# for wealth_segment
New_customer['wealth_segment'].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [97]:
# for gender
New_customer.gender.value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [98]:
# for country 
New_customer.country.value_counts()

Australia    1000
Name: country, dtype: int64

In [99]:
# for state
New_customer.state.value_counts()

NSW    506
VIC    266
QLD    228
Name: state, dtype: int64

In [100]:
# for address
New_customer.address.value_counts()

45 Shopko Center           1
6771 Pleasure Terrace      1
46 Westerfield Place       1
75024 Ronald Regan Hill    1
15 Fisk Road               1
                          ..
00 Judy Terrace            1
79 Armistice Junction      1
7 Clemons Circle           1
6455 Mayfield Street       1
607 Memorial Avenue        1
Name: address, Length: 1000, dtype: int64

In [101]:
# for owns_car
New_customer.owns_car.value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

### Dropping unnecessary columns

In [102]:
# dropping insignificant columns
New_customer.drop(['job_title','deceased_indicator'],inplace = True, axis = 1)

In [103]:
# dropping unknown columns
New_customer.drop(['Unnamed: 16','Unnamed: 17','Unnamed: 18','Unnamed: 19','Unnamed: 20'],inplace = True,axis = 1)

### Adding Age column

In [104]:
New_customer['DOB'] = pd.to_datetime(New_customer['DOB'])
New_customer['Age'] = 2017 - New_customer['DOB'].dt.year
New_customer['Age']

0      60.0
1      47.0
2      43.0
3      38.0
4      52.0
       ... 
959    50.0
971    18.0
975    49.0
980    25.0
983    66.0
Name: Age, Length: 1000, dtype: float64

### -------------------------------------------------------------------------------------------------------------------------------------------------------------------

### Exporting the data cleaned file in order to use it in other python file for modelling and visulaization purposes 

In [108]:
# create an ExcelWriter object to write to a single Excel file
with pd.ExcelWriter(r'C:\Users\Dell\Downloads\KPMG_Tableau.xlsx') as writer:
    
    # write the transaction data to a sheet named 'Transaction'
    Transaction.to_excel(writer, sheet_name='Transaction', index=False)
    
    # write the CustomerDemographic data to a sheet named 'CustomerDemographic'
    CustomerDemographic.to_excel(writer, sheet_name='CustomerDemographic', index=False)
    
    # write the CustomerAddress data to a sheet named 'CustomerAddress'
    CustomerAddress.to_excel(writer, sheet_name='CustomerAddress', index=False)
    
    # write the New Customer data to a sheet named 'New_customer'
    New_customer.to_excel(writer, sheet_name='NewCustomer', index=False)